File: src/checks/connectivity.ts
import { Check, CheckContext, CheckResult } from '../types.js';

export const connectivityCheck: Check = {
  name: 'Connectivity',
  description: 'Verify API is reachable',
  quick: true,
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();

    // Health endpoint is at /health (root level, not under /v1)
    // Strip /v1 suffix if present to get base URL
    let baseUrl = ctx.apiUrl;
    if (baseUrl.endsWith('/v1')) {
      baseUrl = baseUrl.slice(0, -3);
    } else if (baseUrl.endsWith('/v1/')) {
      baseUrl = baseUrl.slice(0, -4);
    }
    baseUrl = baseUrl.replace(/\/+$/, '');

    try {
      const response = await fetch(`${baseUrl}/health`);
      const duration = performance.now() - start;

      if (response.ok) {
        return {
          name: 'Connectivity',
          success: true,
          duration,
          message: 'API reachable',
        };
      } else {
        return {
          name: 'Connectivity',
          success: false,
          duration,
          message: 'API not responding',
          suggestion: `Check if the Drip backend is running at ${baseUrl}`,
        };
      }
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as Error;
      return {
        name: 'Connectivity',
        success: false,
        duration,
        message: 'Connection failed',
        details: err.message,
        suggestion: `Verify DRIP_API_URL (${ctx.apiUrl}) is correct and the server is running`,
      };
    }
  },
};

export const authenticationCheck: Check = {
  name: 'Authentication',
  description: 'Check API key validity',
  quick: true,
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();

    try {
      // Use /customers endpoint with limit=1 to verify API key
      // The /auth/verify endpoint is for Privy tokens (dashboard), not API keys
      const response = await fetch(`${ctx.apiUrl}/customers?limit=1`, {
        headers: { 'Authorization': `Bearer ${ctx.apiKey}` },
      });
      const duration = performance.now() - start;

      if (response.ok) {
        return {
          name: 'Authentication',
          success: true,
          duration,
          message: 'API key valid',
        };
      } else if (response.status === 401 || response.status === 403) {
        return {
          name: 'Authentication',
          success: false,
          duration,
          message: 'Invalid API key',
          suggestion: 'Check your DRIP_API_KEY environment variable',
        };
      } else {
        // Other errors (500, etc.) - key may still be valid
        return {
          name: 'Authentication',
          success: true,
          duration,
          message: 'API key accepted',
          details: `Server returned ${response.status}`,
        };
      }
    } catch (error) {
      const duration = performance.now() - start;
      return {
        name: 'Authentication',
        success: false,
        duration,
        message: 'Auth check failed',
        details: (error as Error).message,
      };
    }
  },
};

File: python/checks/connectivity.py
"""Connectivity and authentication checks."""
from ..types import Check, CheckContext, CheckResult
from ..drip_client import create_client


async def _connectivity_check(ctx: CheckContext) -> CheckResult:
    """Verify SDK can reach the API."""
    try:
        client = create_client(ctx.api_key, ctx.api_url)

        # Try to ping or make a basic request
        if hasattr(client, 'ping'):
            result = client.ping()
            return CheckResult(
                name="connectivity",
                success=True,
                duration=0,
                message="API is reachable",
                details=f"Response: {result}"
            )
        else:
            # Fallback: try listing customers with limit 1
            client.list_customers(limit=1)
            return CheckResult(
                name="connectivity",
                success=True,
                duration=0,
                message="API is reachable",
                details="Connected via list_customers"
            )
    except Exception as e:
        return CheckResult(
            name="connectivity",
            success=False,
            duration=0,
            message=f"Cannot reach API: {e}",
            suggestion="Check DRIP_API_URL and network connectivity"
        )


connectivity_check = Check(
    name="connectivity",
    description="Verify API connectivity",
    run=_connectivity_check,
    quick=True
)


async def _authentication_check(ctx: CheckContext) -> CheckResult:
    """Verify API key is valid."""
    try:
        client = create_client(ctx.api_key, ctx.api_url)
        # Make an authenticated request
        client.list_customers(limit=1)
        return CheckResult(
            name="authentication",
            success=True,
            duration=0,
            message="API key is valid"
        )
    except Exception as e:
        error_str = str(e).lower()
        if "authentication" in error_str or "unauthorized" in error_str or "401" in error_str:
            return CheckResult(
                name="authentication",
                success=False,
                duration=0,
                message="Invalid API key",
                suggestion="Check DRIP_API_KEY environment variable"
            )
        return CheckResult(
            name="authentication",
            success=False,
            duration=0,
            message=f"Authentication check failed: {e}"
        )


authentication_check = Check(
    name="authentication",
    description="Verify API key authentication",
    run=_authentication_check,
    quick=True
)
